src/lib.rs

```diff
@@ -1,7 +1,10 @@
-#[cfg(test)]
-mod tests {
-    #[test]
-    fn it_works() {
-        assert_eq!(2 + 2, 4);
-    }
+#![allow(dead_code)]
+use libc::c_void;
+
+type Account = *mut c_void;
+
+extern "C" {
+	fn create_account() -> Account;
+	fn print(account: Account);
+	fn destroy_account(account: Account);
```

Comment:
What manages the lifetime of the account data? The plugins or the core? I was thinking core, but adding "destroy" makes it seem like the plugin would do it.

Reply:
The plugins would, the reason why is because the core won't know what's going on with the Account, and thus cannot be relied on to destroy it.

The plugin that created it should know how to clean it up, so that's why we're having Plugin be responsible for cleaning up the data.

That doesn't mean that in core we can't have a destruct or for a `Plugin` class that calls the corresponding `destroy_account` function when the plugin needs to be cleared out, but from core's perspective it doesn't make sense to handle the lifetime of the plugin.

Reply:
As an example, say we have a C++ plugin that defines the following as it's account

```cpp
class Account {
private:
    int *thing;
public:
    Account() { thing = new int[5](); }
    ~Account() { delete[] thing; }
};
```

If the core manages the lifetime of the account object, it can only free up the pointer, leaving the allocated `int *` around, thus leaking memory.

Reply:
I think with the exception of some special cases, it may make sense for the core to manage the lifetime. Unless you want to manage both an object by the plugin and by the core. Because I think the core would deal with the archival and loading of old of messages and stuff.
In my design, the accounts would be deleted once the plugin objects are deleted.

And once the API is ready for it, "create_account" would be load_account, with the UUID passed in.

Reply:
Nothing with this design says that core can't archive old messages...

Here's the general plan for this setup

In the core we'll have an IMAccount struct, this contains the Account pointer from the plugin and a reference to the plugin that loaded it.

The core ultimately calls the `destroy_account` function, and then the plugin cleans itself out

Reply:
Okay. Is there anything identifying any specific account in this section of the API? Like are we constructing just an account, or are we constructing a specific account?

Reply:
Currently nothing identifying yet, we can have a discussion on Account identification later, when we get to that point.